ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +160,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&rayv1alpha1.RayService{}).
 		Complete(r)
 }
+
+func (r *RayServiceReconciler) getRayServiceInstance(ctx context.Context, request ctrl.Request) (*rayv1alpha1.RayService, error) {
+	rayServiceInstance := &rayv1alpha1.RayService{}
+	if err := r.Get(ctx, request.NamespacedName, rayServiceInstance); err != nil {
+		if errors.IsNotFound(err) {
+			rayServiceLog.Info("Read request instance not found error!")
+		} else {
+			rayServiceLog.Error(err, "Read request instance error!")
+		}
+		// Error reading the object - requeue the request.
+		return nil, err
+	}
+	return rayServiceInstance, nil
+}
+
+func (r *RayServiceReconciler) updateState(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService, status rayv1alpha1.ServiceStatus, err error) error {
+	rayServiceInstance.Status.ServiceStatus = status
+	if errStatus := r.Status().Update(ctx, rayServiceInstance); errStatus != nil {
+		return fmtErrors.Errorf("combined error: %v %v", err, errStatus)
+	}
+	return err
+}
+
+func (r *RayServiceReconciler) getOrCreateRayClusterInstance(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	rayClusterNamespacedName := types.NamespacedName{
+		Namespace: rayServiceInstance.Namespace,
+		Name:      rayServiceInstance.Name + rayClusterSuffix,
```

Comment:
I feel we should probably generate a uuid to append to the name here, to avoid collisions of resources between a terminating RayCluster and its replacement.